async/timeout.ts

```diff
@@ -0,0 +1,18 @@
+import { deferred } from "./deferred.ts";
+
+export class TimeoutError extends Error {
```

Comment:
I don't think we need a concrete error type here, a simple `Error` would do just fine.

Reply:
Well... Then how can we distinguish if an error occurred from an original promise or `timeout` itself?
I think the `timeout` adds extra context which is not a part of the core concern of the original promise so users would want to distinguish the error.